File: runtimes/swift/Sources/Twilic/Core/Wire.swift
import Foundation

enum Wire {
    static func encodeVaruint(_ value: UInt64, _ out: inout Data) {
        var v = value
        if v < 0x80 {
            out.append(UInt8(v))
            return
        }
        while true {
            var b = UInt8(v & 0x7F)
            v >>= 7
            if v != 0 { b |= 0x80 }
            out.append(b)
            if v == 0 { break }
        }
    }

    static func encodeZigzag(_ value: Int64) -> UInt64 {
        UInt64(bitPattern: (value << 1) ^ (value >> 63))
    }

    static func decodeZigzag(_ value: UInt64) -> Int64 {
        Int64(bitPattern: (value >> 1) ^ UInt64(bitPattern: -(Int64(value & 1))))
    }

    static func encodeBytes(_ bytes: Data, _ out: inout Data) {
        encodeVaruint(UInt64(bytes.count), &out)
        out.append(bytes)
    }

    static func encodeString(_ value: String, _ out: inout Data) {
        encodeBytes(Data(value.utf8), &out)
    }

    static func encodeBitmap(_ bits: [Bool], _ out: inout Data) {
        encodeVaruint(UInt64(bits.count), &out)
        var current: UInt8 = 0
        for (i, bit) in bits.enumerated() {
            if bit { current |= 1 << (i % 8) }
            if i % 8 == 7 {
                out.append(current)
                current = 0
            }
        }
        if !bits.isEmpty && bits.count % 8 != 0 {
            out.append(current)
        }
    }

    final class Reader {
        private let input: Data
        private(set) var offset: Int = 0
        private var depth = 0
        private var budget: Int

        init(_ input: Data) {
            self.input = input
            self.budget = min(input.count, 1024) * 1024
        }

        func claimOutput(_ count: Int) throws {
            guard count >= 0, count <= 1 << 20 else { throw TwilicErrors.invalidData("decode count limit exceeded") }
            guard count <= budget / 8 else { throw TwilicErrors.invalidData("decode output ratio exceeded") }
            budget -= count * 8
        }
        func readIndex() throws -> Int {
            guard let n = Int(exactly: try readVaruint()) else { throw TwilicErrors.invalidData("decode length overflow") }
            return n
        }
        func readCount(_ maximum: Int = 1 << 20) throws -> Int {
            let count = try readIndex()
            guard count <= maximum else { throw TwilicErrors.invalidData("decode count limit exceeded") }
            try claimOutput(count)
            return count
        }
        func enterDepth() throws {
            guard depth < 64 else { throw TwilicErrors.invalidData("decode depth limit exceeded") }
            depth += 1
        }
        func leaveDepth() { depth -= 1 }

        var position: Int { offset }
        var isEOF: Bool { offset >= input.count }

        func readU8() throws -> UInt8 {
            guard offset < input.count else { throw TwilicErrors.unexpectedEOF() }
            let b = input[offset]
            offset += 1
            return b
        }

        func readExact(_ n: Int) throws -> Data {
            guard n >= 0, n <= input.count - offset else { throw TwilicErrors.unexpectedEOF() }
            let end = offset + n
            let slice = input.subdata(in: offset ..< end)
            offset = end
            return slice
        }

        func readVaruint() throws -> UInt64 {
            var shift = 0
            var result: UInt64 = 0
            while true {
                if shift >= 64 { throw TwilicErrors.invalidData("varuint too large") }
                let b = try readU8()
                if shift == 63 && b & 0x7E != 0 { throw TwilicErrors.invalidData("varuint too large") }
                result |= UInt64(b & 0x7F) << shift
                if b & 0x80 == 0 { return result }
                shift += 7
            }
        }

        func readI64Zigzag() throws -> Int64 {
            decodeZigzag(try readVaruint())
        }

        func readBytes() throws -> Data {
            let n = try readIndex()
            return try readExact(n)
        }

        func readString() throws -> String {
            let data = try readBytes()
            guard let s = String(data: data, encoding: .utf8) else {
                throw TwilicErrors.utf8Error()
            }
            return s
        }

        func readBitmap() throws -> [Bool] {
            let bitCount = try readCount()
            let byteCount = (bitCount + 7) / 8
            let raw = try readExact(byteCount)
            var bits = [Bool](repeating: false, count: bitCount)
            for i in 0 ..< bitCount {
                bits[i] = ((raw[i / 8] >> (i % 8)) & 1) == 1
            }
            return bits
        }
    }

    static func newReader(_ input: Data) -> Reader { Reader(input) }

    static func u64FromLEBytes(_ b: Data) -> UInt64 {
        precondition(b.count >= 8)
        var value: UInt64 = 0
        for i in 0 ..< 8 {
            value |= UInt64(b[i]) << (8 * i)
        }
        return value
    }

    static func readU64LE(_ reader: Reader) throws -> UInt64 {
        let b = try reader.readExact(8)
        return u64FromLEBytes(b)
    }

    static func readF64LE(_ reader: Reader) throws -> Double {
        Double(bitPattern: try readU64LE(reader))
    }

    static func appendU64LE(_ out: inout Data, _ v: UInt64) {
        out.append(UInt8(v & 0xFF))
        out.append(UInt8((v >> 8) & 0xFF))
        out.append(UInt8((v >> 16) & 0xFF))
        out.append(UInt8((v >> 24) & 0xFF))
        out.append(UInt8((v >> 32) & 0xFF))
        out.append(UInt8((v >> 40) & 0xFF))
        out.append(UInt8((v >> 48) & 0xFF))
        out.append(UInt8((v >> 56) & 0xFF))
    }

    static func appendF64LE(_ out: inout Data, _ v: Double) {
        appendU64LE(&out, v.bitPattern)
    }
}

File: runtimes/go/internal/core/wire.go
package core

import (
	"encoding/binary"
	"math"
	"unicode/utf8"
)

func encodeVaruint(value uint64, out *[]byte) {
	if value < 0x80 {
		*out = append(*out, byte(value))
		return
	}
	for {
		b := byte(value & 0x7F)
		value >>= 7
		if value != 0 {
			b |= 0x80
		}
		*out = append(*out, b)
		if value == 0 {
			break
		}
	}
}

func encodeZigzag(value int64) uint64 {
	return uint64((value << 1) ^ (value >> 63))
}

func decodeZigzag(value uint64) int64 {
	return int64((value >> 1) ^ uint64(-(int64(value & 1))))
}

func encodeBytes(bytes []byte, out *[]byte) {
	encodeVaruint(uint64(len(bytes)), out)
	*out = append(*out, bytes...)
}

func encodeString(value string, out *[]byte) {
	encodeBytes([]byte(value), out)
}

func encodeBitmap(bits []bool, out *[]byte) {
	encodeVaruint(uint64(len(bits)), out)
	var current byte
	for i, bit := range bits {
		if bit {
			current |= 1 << (i % 8)
		}
		if i%8 == 7 {
			*out = append(*out, current)
			current = 0
		}
	}
	if len(bits)%8 != 0 {
		*out = append(*out, current)
	}
}

type Reader struct {
	input  []byte
	offset int
	depth  int
	budget uint64
}

func newReader(input []byte) *Reader {
	return &Reader{input: input, budget: uint64(min(len(input), 1024)) * 1024}
}

func (r *Reader) claimOutput(count uint64) error {
	if count > 1<<20 {
		return invalidData("decode count limit exceeded")
	}
	if count > r.budget/8 {
		return invalidData("decode output ratio exceeded")
	}
	r.budget -= count * 8
	return nil
}

func (r *Reader) readCount(maximum ...uint64) (uint64, error) {
	n, err := r.readVaruint()
	if err != nil {
		return 0, err
	}
	max := uint64(1 << 20)
	if len(maximum) > 0 {
		max = maximum[0]
	}
	if n > max {
		return 0, invalidData("decode count limit exceeded")
	}
	if err := r.claimOutput(n); err != nil {
		return 0, err
	}
	return n, nil
}

func (r *Reader) enterDepth() error {
	if r.depth >= 64 {
		return invalidData("decode depth limit exceeded")
	}
	r.depth++
	return nil
}
func (r *Reader) leaveDepth() { r.depth-- }

func (r *Reader) position() int {
	return r.offset
}

func (r *Reader) isEOF() bool {
	return r.offset >= len(r.input)
}

func (r *Reader) readU8() (byte, error) {
	if r.offset >= len(r.input) {
		return 0, unexpectedEOF()
	}
	b := r.input[r.offset]
	r.offset++
	return b, nil
}

func (r *Reader) readExact(n int) ([]byte, error) {
	end := r.offset + n
	if n < 0 || n > len(r.input)-r.offset {
		return nil, unexpectedEOF()
	}
	slice := r.input[r.offset:end]
	r.offset = end
	return slice, nil
}

func (r *Reader) readVaruint() (uint64, error) {
	start := r.offset
	var shift uint32
	var result uint64
	for {
		if shift >= 64 {
			return 0, invalidData("varuint too large")
		}
		b, err := r.readU8()
		if err != nil {
			return 0, err
		}
		if shift == 63 && (b&0x7E) != 0 {
			return 0, invalidData("varuint too large")
		}
		result |= uint64(b&0x7F) << shift
		if b&0x80 == 0 {
			if r.offset-start != varuintEncodedLen(result) {
				return 0, invalidData("varuint overlong")
			}
			return result, nil
		}
		shift += 7
	}
}

func varuintEncodedLen(value uint64) int {
	if value == 0 {
		return 1
	}
	length := 0
	for value > 0 {
		length++
		value >>= 7
	}
	return length
}

func (r *Reader) readI64Zigzag() (int64, error) {
	encoded, err := r.readVaruint()
	if err != nil {
		return 0, err
	}
	return decodeZigzag(encoded), nil
}

func (r *Reader) readBytes() ([]byte, error) {
	n, err := r.readVaruint()
	if err != nil {
		return nil, err
	}
	if n > uint64(len(r.input)-r.offset) {
		return nil, unexpectedEOF()
	}
	return r.readExact(int(n))
}

func (r *Reader) readString() (string, error) {
	n, err := r.readVaruint()
	if err != nil {
		return "", err
	}
	if n > uint64(len(r.input)-r.offset) {
		return "", unexpectedEOF()
	}
	bytes, err := r.readExact(int(n))
	if err != nil {
		return "", err
	}
	if !utf8.Valid(bytes) {
		return "", utf8Error()
	}
	return string(bytes), nil
}

func (r *Reader) readBitmap() ([]bool, error) {
	bitCount, err := r.readCount()
	if err != nil {
		return nil, err
	}
	byteCount := int((bitCount + 7) / 8)
	bytes, err := r.readExact(byteCount)
	if err != nil {
		return nil, err
	}
	bits := make([]bool, bitCount)
	for i := 0; i < int(bitCount); i++ {
		bits[i] = ((bytes[i/8] >> (i % 8)) & 1) == 1
	}
	return bits, nil
}

func readU64LE(r *Reader) (uint64, error) {
	b, err := r.readExact(8)
	if err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint64(b), nil
}

func readF64LE(r *Reader) (float64, error) {
	u, err := readU64LE(r)
	if err != nil {
		return 0, err
	}
	return math.Float64frombits(u), nil
}

func appendU64LE(out *[]byte, v uint64) {
	var buf [8]byte
	binary.LittleEndian.PutUint64(buf[:], v)
	*out = append(*out, buf[:]...)
}

func appendF64LE(out *[]byte, v float64) {
	appendU64LE(out, math.Float64bits(v))
}
